### PR TITLE
[GridBundle] Remove unsupported drivers during container compilation

### PR DESCRIPTION
**Pull request by dantleech**
| Q               | A
| --------------- | ---
| Bug fix?        | -
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | #4809 
| License         | MIT

This PR allows the drivers in the GridBundle to be explicitly enable.

The default behavior is to enable both dbal and orm drivers:

```yaml
sylius_grid:
    drivers: [ "doctrine/orm", "doctrine/dbal" ]
```

### CODE REVIEW

File: src/Sylius/Bundle/GridBundle/DependencyInjection/Configuration.php [hunk 1]
```diff
@@ -38,6 +39,20 @@ public function getConfigTreeBuilder()
     /**
      * @param ArrayNodeDefinition $node
      */
+    private function addDriversSection(ArrayNodeDefinition $node)
+    {
+        $node
+            ->children()
+                ->arrayNode('drivers')
```
Comment (dantleech) [inline src/Sylius/Bundle/GridBundle/DependencyInjection/Configuration.php]: Not sure how ambiguous this is -- I assume the user could equally register their own drivers, in whch case they would not need to include them here.

Should add an `->info` element.

**Comment by dantleech** [comment]
Or maybe it would be better to use either `class_exists` or a compiler pass to activate the drivers.

**Comment by dantleech** [comment]
@pjedrzejewski do you think this would be best done as a compiler pass? e.g. if the `doctrine.orm.entuty_manager` exists, then add the `doctrine/orm` driver?

**Comment by pjedrzejewski** [comment]
There is a chance we get into trouble with compiler pass, but that would be nicer DX indeed. Let's try it!

**Comment by dantleech** [comment]
Updated. I updated the `RegisterDriverPass` to check the dependencies/references of the driver definitions, if the reference is not available in the container, then the driver will be removed (and not added to the registry).

This is still slightly bad DX as it hides the avialability of the drivers - 

An alternative could be to replace the invalid definitions with an `UnsupportedDriver` class which would throw an exception upon instantiation asking the user to install the required dependencies (e.g. doctrine ORM, PHPCR ODM). But I feel that might be over-engineering for this specific bundle.

**Comment by pjedrzejewski** [comment]
I agree that removing the services without notice to the developer can be a bit confusing. If someone adds a custom driver and make a typo or reference wrong service, his driver will not appear. So maybe more explicit would be your initial solution where you define enabled drivers in yaml. At least for the ones supplied by SyliusGridBundle default, all custom drivers should be loaded automatically. I'd just split ``drivers.xml`` and put doctrine/orm and doctrine/dbal in separate files and load them or not based on this configuration, wdyt?

**Comment by pamil** [comment]
In the end, I think the best solution for both ResourceBundle and GridBundle drivers is to make the drivers like [security authentication listeners in Symfony SecurityBundle](http://symfony.com/doc/current/cookbook/security/custom_authentication_provider.html#the-factory) or [configuration sources in Sylius ThemeBundle](http://docs.sylius.org/en/latest/bundles/SyliusThemeBundle/configuration_sources/creating_custom.html). The one I've written for ThemeBundle only modifies the container if it was activated, so you can enforce having some dependencies while there's still possibility to disable problematic driver.

File: src/Sylius/Bundle/GridBundle/Tests/DependencyInjection/Compiler/RegisterDriversPassTest.php [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+namespace Sylius\Bundle\GridBundle\Tests\DependencyInjection;
+
+use Symfony\Component\DependencyInjection\Container;
+use Sylius\Bundle\GridBundle\DependencyInjection\Compiler\RegisterDriversPass;
+use Symfony\Component\DependencyInjection\Definition;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Reference;
+
+class RegisterDriversPassTest extends \PHPUnit_Framework_TestCase
```
Comment (pamil) [inline src/Sylius/Bundle/GridBundle/Tests/DependencyInjection/Compiler/RegisterDriversPassTest.php]: This lib makes testing compiler passes & extensions a lot easier - https://github.com/matthiasnoback/SymfonyDependencyInjectionTest#testing-a-compiler-pass :)

**Comment by pjedrzejewski** [comment]
@pamil Yeah, that would be the best, but will require a lot of work now, with not that big benefit currently. Let's start with something simpler to implement. :)

File: src/Sylius/Bundle/GridBundle/Tests/DependencyInjection/Compiler/RegisterDriversPassTest.php (hunk 2 above)
Comment (dantleech) [inline src/Sylius/Bundle/GridBundle/Tests/DependencyInjection/Compiler/RegisterDriversPassTest.php]: tbh, I used to use that library myself but over time I (personally) found that it was just obfuscating the tests. imo it's not worth abstracting.

**Comment by dantleech** [comment]
I also like the "provider" approach - it means that all the "drivers" will be available (i.e. if the user makes a typo it means that we can show them all the correct "answers" - good discoverability) and the providers can also throw good exceptions (i.e. "You need to composer require "doctrine/phpcr-odm") in order to use the `doctrine/phpcr-odm` driver.

On the other hand, by explicitly enabling the "drivers" we oblige the user to do this for every single bundle which has drivers - which is not such good DX.

But also don't mind implementing the explicit solution for now.

**Comment by dantleech** [comment]
Replaced by https://github.com/Sylius/Sylius/pull/4972

**Pull request closed without merging**